src/Package/Impl/Shell/Services/VsOutputService.cs

```diff
@@ -0,0 +1,51 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Common.Core;
+using Microsoft.Common.Core.Logging;
+using Microsoft.Common.Core.Services;
+using Microsoft.Common.Core.Threading;
+using Microsoft.R.Common.Core.Output;
+using Microsoft.VisualStudio.R.Package.Logging;
+
+namespace Microsoft.VisualStudio.R.Package.Shell {
+    internal sealed class VsOutputService : IOutputService {
+        private readonly IServiceContainer _services;
+        private readonly Dictionary<string, IOutput> _outputs;
+        private IMainThread _mainThread;
+
+        public VsOutputService(IServiceContainer services) {
+            _services = services;
+            _mainThread = services.MainThread();
+            _outputs = new Dictionary<string, IOutput>();
+        }
+
+        public async Task<IOutput> GetAsync(string name, CancellationToken cancellationToken) {
+            await _mainThread.SwitchToAsync(cancellationToken);
```

Comment:
Does it need to be main thread? `OutputWindowLogWriter` needs main thread only when it needs to show pane for the first time in `EnsurePaneVisible`. Otherwise it is `OutputStringThreadSafe`

Reply:
Agreed.